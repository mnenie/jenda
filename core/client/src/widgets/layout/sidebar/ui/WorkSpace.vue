<script setup lang="ts">
import { computed } from 'vue'
import { Icon } from '@iconify/vue'
import { useRoute } from 'vue-router/auto'
import type { WorkspaceLink } from '@/shared/config/types-shared'
import { useExpanded } from '@/shared/composables'
import { UiBadge, UiButton } from '@/shared/ui'

const props = defineProps<{
  links: WorkspaceLink[]
}>()

const route = useRoute()

function isCurrentPath(link: WorkspaceLink): boolean {
  if (route.name === 'boards') {
    return link.pathName === 'boards'
  }
  if (route.name === 'community') {
    return link.pathName === 'templates'
  }
  return route.path === `/${link.pathName}`
}

const pathName = computed(() => {
  return props.links.map(link => ({
    ...link,
    isActive: isCurrentPath(link),
  }))
})

const expanded = useExpanded()

const { isExpanded } = expanded.getExpanded()

const contentPosition = computed(() => {
  return isExpanded.value ? 'space-between' : 'center'
})
</script>

<template>
  <div class="flex flex-col space-y-1.5 justify-start mb-20px">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: $t(`sidebar.${link.name}`),
        triggers: ['hover'],
        disabled: isExpanded,
      }"
      :to="`/${link.pathName}`"
      class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        size="sm"
        class="w-full gap-2 shadow-none px-2 transition-all duration-200 ease"
        :class="link.isActive ? '!bg-neutral-200/40 dark:!bg-neutral-700/50' : ''"
        :style="{ padding: !isExpanded ? '0px' : '', justifyContent: contentPosition }"
      >
        <div class="flex items-center gap-2">
          <Icon :icon="link.icon" class="!w-15px !h-15px 2xl:(!w-4 !h-4) text-neutral-900 dark:text-neutral-200" />
          <span v-show="isExpanded" class="text-13px 2xl:text-sm !fw500 text-neutral-900 dark:text-neutral-100">{{ $t(`sidebar.${link.name}`) }}</span>
        </div>
        <UiBadge v-if="link.id === 1 && isExpanded" variant="solid" class="px-5px py-0 text-xs rounded-md">
          <span class="text-[10px] text-white">3</span>
        </UiBadge>
      </UiButton>
    </RouterLink>
  </div>
</template>
