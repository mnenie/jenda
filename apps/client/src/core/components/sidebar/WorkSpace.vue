<script setup lang="ts">
import LinkItem from './links/LinkItem.vue'
import type { WorkspaceLink } from '../../types'
import { UiBadge } from '@/shared/ui'

defineProps<{
  links: WorkspaceLink[]
}>()
</script>

<template>
  <div class="flex flex-col space-y-1.5 justify-start mb-20px">
    <LinkItem
      v-for="link in links"
      :key="link.name"
      :link="link"
    >
      <template #badge="{ isExpanded }">
        <UiBadge
          v-if="link.badge && isExpanded && link.name === 'workflows'"
          variant="soft"
          class="px-5px py-0 text-xs rounded-md"
        >
          <span class="text-xs">beta</span>
        </UiBadge>
      </template>
    </LinkItem>
  </div>
</template>
