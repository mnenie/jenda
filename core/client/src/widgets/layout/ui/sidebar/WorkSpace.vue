<script setup lang="ts">
import type { Link } from '../../model'
import { UiButton } from '@/shared/ui'
import { Icon } from '@iconify/vue'
import { computed } from 'vue'
import { useRoute } from 'vue-router'

const props = defineProps<{
  links: Link[]
  isExpanded: boolean
}>()

const route = useRoute()

function isCurrentPath(link: Link) {
  if (route.path === '/boards') {
    return link.pathName === 'boards'
  }
  return route.path === `/${link.pathName}`
}

const pathName = computed(() => {
  return props.links.map(link => ({
    ...link,
    isActive: isCurrentPath(link),
  }))
})

const contentPosition = computed(() => {
  return props.isExpanded ? 'flex-start' : 'center'
})
</script>

<template>
  <div class="flex flex-col gap-5px justify-start mb-30px">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: $t(`sidebar.${link.name}`),
        triggers: ['hover'],
        disabled: isExpanded,
      }"
      :to="link.pathName"
      class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        class="w-full gap-2 shadow-none py-0 px-2 transition-all duration-200 ease"
        :class="link.isActive ? '!bg-neutral-200/40 dark:!bg-neutral-700/50' : ''"
        :style="{ padding: !isExpanded ? '0px' : '', justifyContent: contentPosition }"
      >
        <Icon :icon="link.icon" class="text-17px text-neutral-900 dark:text-neutral-200" />
        <span v-show="isExpanded" class="text-sm !fw500 text-neutral-900 dark:text-neutral-100">{{ $t(`sidebar.${link.name}`) }}</span>
      </UiButton>
    </RouterLink>
  </div>
</template>
