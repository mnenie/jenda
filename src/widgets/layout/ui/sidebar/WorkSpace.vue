<script setup lang="ts">
import { computed } from 'vue';
import { useRoute } from 'vue-router';
import { useI18n } from 'vue-i18n';
import { useDark } from '@vueuse/core';
import { RouteNames } from '@/shared/config/consts';
import { UiButton } from '@/shared/ui';
import type { Link } from '../../types';

const props = defineProps<{
  links: Link[];
  isExpanded: boolean;
}>();

const route = useRoute();
const isDark = useDark();

const { t } = useI18n();

const isCurrentPath = (link: Link) => {
  if (route.name?.toString().startsWith(RouteNames.members)) {
    return link.pathName === RouteNames.members;
  }
  return link.pathName === route.name;
};

const pathName = computed(() => {
  return props.links.map((link) => ({
    ...link,
    isActive: isCurrentPath(link)
  }));
});

const contentPosition = computed(() => {
  return props.isExpanded ? 'flex-start' : 'center';
});
const iconColor = computed(() => {
  if (isDark.value) {
    return !props.isExpanded ? 'var(--zinc-200)' : 'var(--zinc-300)';
  } else {
    return !props.isExpanded ? 'var(--zinc-800)' : 'rgb(82 82 91 / 0.9)';
  }
});
</script>

<template>
  <p v-show="isExpanded" :class="[$style.name_block, 'text-sm']">{{ t('sidebar.section') }}</p>
  <div :class="$style.sidebar_main_links">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: t(`sidebar.${link.title}`),
        triggers: ['hover'],
        disabled: isExpanded
      }"
      :to="{ name: link.pathName }"
      :class="$style.link"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        :class="$style.link_btn"
        :style="{ padding: !isExpanded ? '0px' : '' }"
      >
        <component :is="link.icon" :size="18" :color="iconColor" />
        <span v-show="isExpanded" class="text-sm">{{ t(`sidebar.${link.title}`) }}</span>
      </UiButton>
    </RouterLink>
  </div>
</template>

<style module lang="scss">
@import '@/app/styles/mixins';

.name_block {
  color: var(--zinc-300);
  margin-bottom: 12px;
  text-transform: uppercase;
}

.sidebar_main_links {
  display: flex;
  flex-direction: column;
  gap: 5px;
  justify-content: flex-start;
  margin-bottom: 30px;

  .link {
    display: flex;
    align-items: center;
    gap: 6px;
    cursor: pointer;
    color: var(--zinc-900);
    justify-content: flex-start;
    width: 100%;

    .link_btn {
      width: 100%;
      justify-content: v-bind('contentPosition');
      gap: 6px;
      box-shadow: none;
    }

    & span {
      font-weight: 500 !important;
      color: var(--zinc-900);
    }
  }
}

:global(html.dark) {
  .sidebar_main_links {
    .link {
      .link_btn {
        & span {
          color: var(--zinc-100);
        }
      }
    }
  }
}
</style>
