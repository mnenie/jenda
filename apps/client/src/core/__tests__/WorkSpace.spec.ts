import { mount, RouterLinkStub } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { useRoute } from 'vue-router/auto'
import WorkSpace from '../components/sidebar/WorkSpace.vue'

vi.mock('vue-router/auto')

vi.mock('@/shared/composables/expanded', () => ({
  useExpandedContext: vi.fn(() => ({
    isExpanded: {
      value: true,
    },
  })),
}))

describe('tests for WorkSpace.vue', () => {
  // @ts-expect-error mock types
  useRoute.mockReturnValue({ name: 'boards' })
  const wrapper = mount(WorkSpace, {
    global: {
      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'sidebar.boards': 'Boards',
            'sidebar.section': 'Workspace',
          }
          return translations[key]
        },
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      directives: {
        tooltip() {},
      },
    },
    props: {
      links: [
        {
          id: 0,
          name: 'boards',
          pathName: 'boards',
          icon: 'i-hugeicons-trello',
        },
      ],
    },
  })

  it('should be render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be redirect with RouterLink correctly', () => {
    expect(wrapper.findComponent(RouterLinkStub).props('to')).toEqual({
      name: 'boards',
    })
  })
})
