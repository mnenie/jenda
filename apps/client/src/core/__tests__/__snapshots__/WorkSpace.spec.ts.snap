// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for WorkSpace.vue > should be render correctly 1`] = `
"<div class="flex flex-col space-y-1.5 justify-start mb-20px"><a class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"><button class="btn btn-secondary h-9 2xl:h-8 rounded-md w-full gap-2 shadow-none px-2 transition-all duration-200 ease !bg-neutral-200/40 dark:!bg-neutral-700/50" style="justify-content: space-between;">
      <!--v-if-->
      <div class="flex items-center gap-2"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" role="img" class="!w-16px !h-16px !2xl:w-4 !2xl:h-4 text-neutral-900 dark:text-neutral-200" width="1em" height="1em" viewBox="0 0 16 16"></svg><span class="text-default !fw500 text-neutral-900 dark:text-neutral-100">Boards</span></div>
      <!--v-if-->
    </button></a></div>"
`;
